### ISSUE TITLE
Not able to filter on Array

### ISSUE THREAD

**Issue by mas222r** [bohnman/squiggly-filter-jackson#18]
Hi,

I am struggling to retrieve the Array elements in a given json. Could you please help me in resolving this.

for an example i have below json:

{
    "store": {
        "book": [
            {
                "category": "reference",
                "author": "Nigel Rees",
                "title": "Sayings of the Century",
                "price": 8.95
            },
            {
                "category": "fiction",
                "author": "Evelyn Waugh",
                "title": "Sword of Honour",
                "price": 12.99
            },
            {
                "category": "fiction",
                "author": "Herman Melville",
                "title": "Moby Dick",
                "isbn": "0-553-21311-3",
                "price": 8.99
            },
            {
                "category": "fiction",
                "author": "J. R. R. Tolkien",
                "title": "The Lord of the Rings",
                "isbn": "0-395-19395-8",
                "price": 22.99
            }
        ],
        "bicycle": {
            "color": "red",
            "price": 19.95
        }
    },
    "expensive": 10,
	"issueId":{
	   "issueIdType":"history",
	   "issueIdValue":"found"
	},
	"issueId2":{
	   "issueIdType2":"history",
	   "issueIdValue2":"found",
	   "issueSubCategory":{
	         "issueSubCategoryType":"global",
			 "issueSubCategoryValue":"more Found"	 
	   },
	   "issueSearchArray":[{
	         "issueSearchArrayType":"global",
			 "issueSearchArrayTypeValue":"more Found"	 
	   }]
	}
}

My Scenarios are:
1. As i have book array i need to get Specific book array element like book[0] only or book[1] or book[2] or book[3]
2. I need to retrieve selected book array elements from 0 to 2 or 1 to 2 or 1 to 3 or 2 to 3.

**Comment by bohnman** [comment]
Filtering arrays by their indices is not possible today.  Squiggly is implemented as a Jackson PropertyFilter which doesn't expose array index information.

However, I will investigate to see if there's a way to make this possible for a future version of Squiggly.

**Comment by spartabantam** [comment]
That would be incredibly useful

**Comment by passionate-developer** [comment]
yep, it is a great library bohnman. It will be very useful if we can have arrays implementation for this library as posted by mas222r user above. 

Do you have any plans when the new release will be available with this implementation? Thanks

**Comment by bohnman** [comment]
Did some investigation into a solution last night.

The Good News:  It is possible!  Yay!!!!
The Bad News:  It involves using reflection monkey patch some Jackson code to make it work, which means it's more susceptible to compatibility changes.

As far as timeline, I'm thinking within a month, but don't hold me too that.  I have several ideas for a 2.0 version of Squiggly and this will most likely be part of the 2.0 release.

**Comment by bohnman** [comment]
I created a more general issue that should handle this use case:  #19

**Comment by mas222r** [comment]
Million thanks :) . This will reduce my lengthy code now.

**Comment by bohnman** [comment]
UPDATE: I was finally had some time over the weekend to work on Squiggly 2.0.  I have most of the features completed.  I expect a release in 1 to 2 weeks.  I'll keep you posted!

**Comment by passionate-developer** [comment]
Great news. I will wait for the release package. Hope the inside array fix is resolved :) - Thanks

**Comment by bohnman** [comment]
@passionate-developer Are you referring to issue #29? If so, I already have fix in master and can release it if you'd like it now.

**Comment by passionate-developer** [comment]
I was referring to issue 18. I was having issue where both exclude and include for array indexes were not working e.g I couldn't do array[n] to give me just n element or -array [n] to exclude only array [n] element as described by mas222r user above.

**Comment by bohnman** [comment]
@passionate-developer -  In 2.0, the way to manipulate arrays will be through the use of functions.

There will be a number of built-in functions.  For example, to get a simple subset of an array/list, you will be able to use the slice function:

``fields=friends.slice(0, 2)``

In the case above, you would get a new array, with element 0 and element 1

There most likely be a more complicated version of array slicing, called pick where you can get arbitrary cross-sections of an array/list.  For example:

``fields=friends.pick(1, 2, [5:-1])``

In the case above, you would get a new array with element 1, element 2, and elements 5 through the end of the array.

In addition, you will be able to register new functions or override built-in functions.

For example:

```java
// Registers all public static methods of a utility class as functions
Squiggly.builder().functions(SquigglyFunctions.create(MyUtilityClass.class)).build();
```

**Comment by passionate-developer** [comment]
Excellent.  Just wondering if you know if your squiggly library code is used by any communities or other companies as well.  Cheers

**Comment by bohnman** [comment]
@passionate-developer That's a good question.  I am not sure what companies are using this library other than the one I work for , http://www.biohitech.com

**Comment by Calvin-J-Scott** [comment]
@bohnman Hey mate just wondered if there was any update on how you was getting on with this?

**Comment by passionate-developer** [comment]
Hi is there any limit on how many json fields can be included or excluded? Apparently it failed to respond after 62 fields were requested in include.

**Comment by bohnman** [comment]
@passionate-developer There is not a limit set.   Could you provide more detail?  Thanks!

**Comment by mas222r** [comment]
@bohnman  is the manipulation of arrays is part of latest version?? if yes could you please elaborate how to use it .

I have used like below as per your above comments but got exception 

includes = companyList.companySubGroupNames.companySubGrpGivenNames.slice(0,2)

java.lang.IllegalArgumentException: com.fasterxml.jackson.databind.JsonMappingException: line 1:62 extraneous input '(' expecting {<EOF>, ',', '.'} (through reference chain: java.util.LinkedHashMap["companyList"])

**Comment by Rgkol** [comment]
@bohnman 
Hello, I guess the version 2.0 with the array manipulation was never released?
Or am I missing something?

**Comment by manish-bharti-cashify** [comment]
@bohnman 
Hello. Is the version with Array manipulation released?
I am not able to find the slice or pick function.
